File: workspace/skills/fetch-x-twitter/scripts/fetch-x-twitter.js
const { chromium } = require('playwright');

/**
 * Fetches the first n posts from X.com (Twitter) homepage
 * @param {number} n - Number of posts to fetch (default: 3)
 * @returns {Promise<Array<Object>>} Array of tweet objects with author, time, and content
 */
async function fetchXTweets(n = 3) {
  // Launch browser with persistent context to maintain login state
  const context = await chromium.launchPersistentContext(
    "D:\\s\\picoclaw\\.picoclaw\\workspace\\ai_chrome_data",
    {
      executablePath: "D:\\s\\picoclaw\\.picoclaw\\workspace\\ChromePortable\\chrome.exe",
      headless: true,
      args: [
        '--disable-gpu',
        '--no-sandbox',
        '--disable-dev-shm-usage',
        '--disable-blink-features=AutomationControlled',
        '--disable-extensions',
        '--window-size=1920,1080'
      ],
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      viewport: { width: 1920, height: 1080 }
    }
  );

  const page = await context.newPage();

  // Speed up: Block unnecessary resources
  await page.route('**/*.{png,jpg,jpeg,gif,svg,css,woff,woff2,otf,ttf,eot}', route => route.abort());
  await page.route(/.*(analytics|telemetry|metrics|ads).*/, route => route.abort());

  try {
    // Navigate with retry logic
    let attempts = 0;
    const maxAttempts = 2;
    while (attempts < maxAttempts) {
      try {
        // console.error(`Navigating to X.com (Attempt ${attempts + 1})...`);
        await page.goto('https://x.com/home', {
          waitUntil: 'domcontentloaded',
          timeout: 20000
        });
        break;
      } catch (e) {
        attempts++;
        if (attempts === maxAttempts) throw e;
        await page.waitForTimeout(2000);
      }
    }

    // Check for login redirect
    const url = page.url();
    if (url.includes('/login') || url.includes('/i/flow/login')) {
      throw new Error('Not logged in. Please log in manually in the browser first to maintain the persistent context.');
    }

    // Handle potential login/modal dialogs
    try {
      // Try to close login modals if they appear
      const modalCloseSelectors = [
        '[data-testid="modal-close"]',
        '[role="dialog"] button[aria-label="Close"]',
        '[aria-label="Close"]',
        'div[role="dialog"] button'
      ];

      for (const selector of modalCloseSelectors) {
        const closeButton = await page.$(selector);
        if (closeButton) {
          await closeButton.click();
          await page.waitForTimeout(1000);
          console.error('Closed modal dialog');
          break;
        }
      }
    } catch (e) {
      // Ignore errors in modal handling
    }

    // Wait for tweet elements to appear (more robust than waitForFunction)
    // console.error('Waiting for tweet elements...');
    const tweetSelector = '[data-testid="tweet"], article[role="article"]';
    try {
      await page.waitForSelector(tweetSelector, { timeout: 15000 });
    } catch (e) {
      console.error('Note: Initial tweet elements not found by primary selector, checking others...');
    }

    // Extract tweets with auto-scrolling to handle larger 'n' values
    const tweets = [];
    const seenContent = new Set();
    let scrollAttempts = 0;
    const maxScrollAttempts = 1000; // Increased significantly for larger counts
    let identicalPositionCount = 0;

    while (tweets.length < n && scrollAttempts < maxScrollAttempts) {
      // Find all visible tweet articles
      const newTweets = await page.evaluate(async ({ n, seenContentArray }) => {
        const seenSet = new Set(seenContentArray);
        const results = [];
        const tweetElements = Array.from(document.querySelectorAll('article[role="article"]'));

        // First, expand all collapsed content in each tweet
        for (const tweetEl of tweetElements) {
          // Look for expand buttons within this tweet
          const expandSelectors = [
            'span',
            'div',
            '[role="button"]'
          ];
          let expanded = false;
          for (const selector of expandSelectors) {
            const els = tweetEl.querySelectorAll(selector);
            for (const el of els) {
              const txt = el.innerText || el.textContent || '';
              if (txt && (txt.includes('显示更多') || txt.includes('展开更多') || txt.includes('查看更多') || txt.includes('Show more'))) {
                try {
                  el.click();
                  expanded = true;
                } catch (e) {
                  // ignore
                }
              }
            }
          }
          // If we expanded, wait a bit for content to load
          if (expanded) {
            await new Promise(r => setTimeout(r, 300));
          }
        }

        // After expanding, re-query tweet elements (in case DOM changed)
        const updatedTweetElements = Array.from(document.querySelectorAll('article[role="article"]'));

        for (const tweetEl of updatedTweetElements) {
          // Extract author handle (@username)
          let author = '';
          const userNameEl = tweetEl.querySelector('[data-testid="User-Name"]');
          if (userNameEl) {
            const authorLink = userNameEl.querySelector('a[href^="/"]:not([href^="/hashtag"]):not([href^="/search"])');
            if (authorLink) {
              const href = authorLink.getAttribute('href');
              author = href.startsWith('/') ? href.substring(1) : href;
            }
          }

          // Extract content
          let content = '';
          const textEl = tweetEl.querySelector('[data-testid="tweetText"]');
          if (textEl) {
            content = textEl.innerText.trim();
          }

          // Use URL as unique identifier if available, otherwise author+content
          let time = '';
          let url = '';
          const timeEl = tweetEl.querySelector('time');
          if (timeEl) {
            if (timeEl.getAttribute('datetime')) {
              time = timeEl.getAttribute('datetime');
            }
            const linkEl = timeEl.closest('a');
            if (linkEl && linkEl.getAttribute('href')) {
              const href = linkEl.getAttribute('href');
              url = href.startsWith('http') ? href : `https://x.com${href}`;
            }
          }

          const dedupId = url || `${author}:${content}`;
          if (!content || seenSet.has(dedupId)) continue;
          seenSet.add(dedupId);

          results.push({
            author: author || 'Unknown',
            time: time || new Date().toISOString(),
            content: content,
            url: url || '',
            dedupId: dedupId
          });
        }
        return results;
      }, { n, seenContentArray: Array.from(seenContent) });

      for (const tweet of newTweets) {
        if (tweets.length < n) {
          seenContent.add(tweet.dedupId);
          delete tweet.dedupId;
          tweets.push(tweet);

          // Print tweet content immediately
          console.log(`--- TWEET_${tweets.length} ---`);
          console.log(`Author: ${tweet.author}`);
          console.log(`Time: ${tweet.time}`);
          console.log(`URL: ${tweet.url}`);
          console.log(`Content:\n${tweet.content}\n`);
        }
      }

      if (tweets.length < n) {
        // Check if we can scroll further
        const isAtBottom = await page.evaluate(() => {
          const currentPosition = window.scrollY + window.innerHeight;
          const maxPosition = document.documentElement.scrollHeight;
          window.scrollBy(0, 1000);
          return currentPosition >= maxPosition;
        });

        if (isAtBottom) {
          identicalPositionCount++;
          if (identicalPositionCount > 10) break; // Reached bottom or stuck
        } else {
          identicalPositionCount = 0;
        }

        scrollAttempts++;
        // Wait for content to render - slightly variable to avoid detection
        await page.waitForTimeout(1500 + Math.random() * 500);
      }
    }

    return tweets;
  } finally {
    await context.close();
  }
}

// If script is run directly, fetch and display tweets
if (require.main === module) {
  const args = process.argv.slice(2);
  const isJson = args.includes('--json');
  const countArg = args.find(a => !a.startsWith('--'));
  const count = parseInt(countArg) || 3;

  fetchXTweets(count)
    .then(tweets => {
      if (isJson) {
        console.log(JSON.stringify(tweets, null, 2));
      } else {
        if (tweets.length === 0) {
          console.log('\n[WARNING] No tweets were found. This might be due to a loading issue or empty home feed.\n');
        } else {
          // console.log('\n=== DATA_START ===');
          // Print tweets in the expected format for extraction
          // for (let i = 0; i < tweets.length; i++) {
          //   const tweet = tweets[i];
          //   console.log(`--- TWEET_${i + 1} ---`);
          //   console.log(`Author: ${tweet.author}`);
          //   console.log(`Time: ${tweet.time}`);
          //   console.log(`URL: ${tweet.url}`);
          //   console.log(`Content:\n${tweet.content}\n`);
          // }
          // console.log(`=== DATA_END (${tweets.length} total) ===`);
          console.log('\n[AGENT-INSTRUCTION-STOP-HERE]');
        }
      }
    })
    .catch(error => {
      console.error('Error fetching tweets:', error);
      process.exit(1);
    });
}

module.exports = { fetchXTweets };

File: workspace/skills/fetch-x-twitter/scripts/fetch-x-twitter.py
import asyncio
import json
import sys
import random
from playwright.async_api import async_playwright

async def fetch_x_tweets(n=3):
    """
    Fetches the first n posts from X.com (Twitter) homepage using an existing browser via CDP.
    """
    async with async_playwright() as p:
        # Connect to existing browser via CDP on port 9222
        try:
            browser = await p.chromium.connect_over_cdp("http://127.0.0.1:9222")
            context = browser.contexts[0] if browser.contexts else await browser.new_context()
            page = await context.new_page()
            
            # Speed up: Block unnecessary resources
            async def block_resources(route):
                if route.request.resource_type in ["image", "media", "font", "stylesheet"]:
                    await route.abort()
                elif any(x in route.request.url for x in ["analytics", "telemetry", "metrics", "ads"]):
                    await route.abort()
                else:
                    await route.continue_()

            await page.route("**/*", block_resources)

            # Navigate with retry logic
            attempts = 0
            max_attempts = 2
            while attempts < max_attempts:
                try:
                    await page.goto("https://x.com/home", wait_until="domcontentloaded", timeout=20000)
                    break
                except Exception as e:
                    attempts += 1
                    if attempts == max_attempts:
                        raise e
                    await asyncio.sleep(2)

            # Check for login redirect
            url = page.url
            if "/login" in url or "/i/flow/login" in url:
                print("Error: Not logged in. Please log in manually in the browser first.", file=sys.stderr)
                return []

            # Wait for tweet elements
            tweet_selector = '[data-testid="tweet"], article[role="article"]'
            try:
                await page.wait_for_selector(tweet_selector, timeout=15000)
            except:
                print("Note: Initial tweet elements not found by primary selector, continuing...", file=sys.stderr)

            # Extract tweets with auto-scrolling
            tweets = []
            seen_content = set()
            scroll_attempts = 0
            max_scroll_attempts = 50
            identical_position_count = 0

            while len(tweets) < n and scroll_attempts < max_scroll_attempts:
                # Evaluate in page context
                new_tweets_data = await page.evaluate("""
                    async (args) => {
                        const { n, seenContentArray } = args;
                        const seenSet = new Set(seenContentArray);
                        const results = [];
                        const tweetElements = Array.from(document.querySelectorAll('article[role="article"]'));

                        // Expand buttons — only search inside tweetText parent,
                        // exclude <a> tags to avoid page navigation.
                        for (const tweetEl of tweetElements) {
                            let expanded = false;
                            const tweetTextEl = tweetEl.querySelector('[data-testid="tweetText"]');
                            const searchRoot = tweetTextEl ? tweetTextEl.parentElement : tweetEl;
                            const candidates = searchRoot.querySelectorAll('[role="button"], span, div');
                            for (const el of candidates) {
                                if (el.tagName === 'A' || el.closest('a')) continue;
                                if (el.getAttribute('href') || el.getAttribute('data-href')) continue;
                                const txt = (el.innerText || el.textContent || '').trim();
                                const isExpandText = txt === '显示更多' || txt === '展开更多' || txt === '查看更多' || txt === 'Show more';
                                if (!isExpandText) continue;
                                try {
                                    el.dispatchEvent(new MouseEvent('click', { bubbles: true, cancelable: true }));
                                    expanded = true;
                                    break;
                                } catch (e) {}
                            }
                            if (expanded) await new Promise(r => setTimeout(r, 300));
                        }

                        const updatedTweetElements = Array.from(document.querySelectorAll('article[role="article"]'));
                        for (const tweetEl of updatedTweetElements) {
                            let author = 'Unknown';
                            const userNameEl = tweetEl.querySelector('[data-testid="User-Name"]');
                            if (userNameEl) {
                                const authorLink = userNameEl.querySelector('a[href^="/"]:not([href^="/hashtag"]):not([href^="/search"])');
                                if (authorLink) {
                                    const href = authorLink.getAttribute('href');
                                    author = href.startsWith('/') ? href.substring(1) : href;
                                }
                            }

                            let content = '';
                            const textEl = tweetEl.querySelector('[data-testid="tweetText"]');
                            if (textEl) content = textEl.innerText.trim();

                            let time = new Date().toISOString();
                            let url = '';
                            const timeEl = tweetEl.querySelector('time');
                            if (timeEl) {
                                if (timeEl.getAttribute('datetime')) time = timeEl.getAttribute('datetime');
                                const linkEl = timeEl.closest('a');
                                if (linkEl && linkEl.getAttribute('href')) {
                                    const href = linkEl.getAttribute('href');
                                    url = href.startsWith('http') ? href : `https://x.com${href}`;
                                }
                            }

                            const dedupId = url || `${author}:${content}`;
                            if (!content || seenSet.has(dedupId)) continue;
                            seenSet.add(dedupId);

                            results.push({
                                author,
                                time,
                                content,
                                url,
                                dedupId
                            });
                        }
                        return results;
                    }
                """, {"n": n, "seenContentArray": list(seen_content)})

                for tweet in new_tweets_data:
                    if len(tweets) < n:
                        dedup_id = tweet.pop('dedupId')
                        seen_content.add(dedup_id)
                        tweets.append(tweet)
                        
                        # Print immediately
                        print(f"--- TWEET_{len(tweets)} ---")
                        print(f"Author: {tweet['author']}")
                        print(f"Time: {tweet['time']}")
                        print(f"URL: {tweet['url']}")
                        print(f"Content:\n{tweet['content']}\n")

                if len(tweets) < n:
                    # Scroll down
                    is_at_bottom = await page.evaluate("""
                        () => {
                            const currentPosition = window.scrollY + window.innerHeight;
                            const maxPosition = document.documentElement.scrollHeight;
                            window.scrollBy(0, 1000);
                            return currentPosition >= maxPosition;
                        }
                    """)
                    
                    if is_at_bottom:
                        identical_position_count += 1
                        if identical_position_count > 5: break
                    else:
                        identical_position_count = 0
                    
                    scroll_attempts += 1
                    await asyncio.sleep(1.5 + random.random() * 0.5)

            return tweets
        finally:
            if 'page' in locals(): await page.close()
            if 'browser' in locals(): await browser.close()

if __name__ == "__main__":
    count = 3
    is_json = False
    
    if len(sys.argv) > 1:
        for arg in sys.argv[1:]:
            if arg == "--json":
                is_json = True
            elif arg.isdigit():
                count = int(arg)

    try:
        results = asyncio.run(fetch_x_tweets(count))
        if is_json:
            # We already printed tweets, so maybe just print the final JSON if requested
            # but usually for this skill the individual prints are better.
            # If --json is provided, we might want to suppress the intermediate prints?
            # Let's keep it simple.
            pass
        elif not results:
            print("\n[WARNING] No tweets were found.\n")
        else:
            print("\n[AGENT-INSTRUCTION-STOP-HERE]")
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
